### ISSUE TITLE
New objective datums not working.

### ISSUE THREAD

**Issue by octareenroon91** [Zaers/-tg-station#174]
I know some typepaths were changed, but this is even a pretty weird set of objectives for a traitor.  Did something get overlooked?

![clipboard01](https://cloud.githubusercontent.com/assets/12420976/10557789/f81ffcd2-7476-11e5-885e-21cb564ac7d9.png)

**Comment by Koriath** [comment]
I can take another look, but I think all the typepaths were changed everywhere, so they should work normally.

**Comment by octareenroon91** [comment]
Nothing, Free Objective, Debrain, and possibly Assume-Identity are all under the "default" branch of objectives.  These need to be excluded from first-choice objectives definitions, especially for traitors.

Also Malf AI is not working well at present.

I need a revert for the Black Ops groups of updates, at the very least.

**Comment by LanCartwright** [comment]
revert it, start from scratch, back to drawing board. I'll call goof up.

**Comment by octareenroon91** [comment]
I am having trouble unraveling https://github.com/Zaers/-tg-station/pull/169 directly.

**Comment by LanCartwright** [comment]
https://github.com/Iamgoofball/-tg-station/tree/need_my_mgs5

**Comment by Iamgoofball** [comment]
Hey, heads up, Debrain just requires having a brain, doesn't need you to do any ling specific shit.

Free Objective is basically any objective that had an error generating the explanation text shit.

I reccomend replacing all Free Objective error text in the code with "Free Objective (OBJECTIVE NAME)" to figure out wtf went wrong

**Comment by octareenroon91** [comment]
I get that, but we were also getting 'steal identity,' which again a traitor COULD do, but probably didn't get before.

**Comment by Iamgoofball** [comment]
that's weird and shouldn't be showing up, that is indeed a ling specific one

**Comment by octareenroon91** [comment]
Yeah, there are instances of ```required_role = "changeling"``` that probably should exist.

**Comment by Iamgoofball** [comment]
yeah, make sure that's slapped on anything that's role specific but still under the default tag.

**Comment by octareenroon91** [comment]
I'm also confused why objectives throw up their hands and go "Free Objective" instead of going for a replacement.

**Comment by Iamgoofball** [comment]
That's something I can add. I'll look into it.

**Comment by octareenroon91** [comment]
I just became aware that adding new assassinate objectives through the TP/antag role management screen doesn't work.  All the prompts still happen, but nothing happens or changes.  There's no feedback implying success or failure: nothing in the admin chat screen nor a runtime in the log.

Further testing suggests all the objectives involving a specific player will fail the same silent way.

**Comment by Koriath** [comment]
Line 664 is: new_objective = new objective_path
Objective path is defined and assigned a few lines before that:
var/objective_path = text2path("/datum/objective/[new_obj_type]")

My guess? the text2path thing needs to be redone, because it doesn't take into accoun the "/default" bit we added in.

**Comment by Iamgoofball** [comment]
so are you guys just gonna completely revert this and recode it from scratch instead of fixing it?

**Comment by octareenroon91** [comment]
The mass revert served only to bring master back to a state that's good to host.

I think we're going to port again, soon, with more care before putting it on master.

**Comment by Iamgoofball** [comment]
okay but here's a reccomendation:

flat out override all custom objectives code with the new code, and then re-add your shit in over it

**Comment by Koriath** [comment]
@Iamgoofball Not my code, I found it like that. I usually prefer not to touch things like gamemodes or any other subsystem that is that import. It's not even the spaghetti, it's the fact that I haven't coded for that long, so I know there's 137% chance of me fucking it up.

That line I found pretty much tries to assign an objective datum to the var, using the string to build the path. Which of course doesn't work since we changed the paths.

My sugestion would be to re-write the way traitor objectives are assigned. And while we're at it, we could do something more interesting than completely randomizing it. I saw some traitor faction stuff for instance, why not start using those? Every faction has a diferent list of objectives it can give players, and players get a prompt to chose the faction they want to work with.
I personally like stealth, so steal and sabotage objectives would be my prefered thing.
Some people like murder, and some like gimmick stuff (waffle group).

**Comment by Iamgoofball** [comment]
@Koriath The string shit appears to be located in admin code, so it's shit neither you or I touched. fucking old-code

and you realize my PR did indeed re-write the way traitor objectives are assigned, right?

and adding choosing objective means people are gonna choose the faction with the easiest objectives and not do anything else

**Comment by Koriath** [comment]
Eh, maybe on /tg/, were we actually have a lot of people who like to try new shit, and a few who like to get challenged. Otherwise, there's a simple solution: the easier factions get shit gear. I mean, they don't need good gear when their objectives are easy. Harder things like "kill the whole crew" or something like that get explosives and guns. "Take Ian for a walk" gets you a bar of soap and a tactical turtleneck.

Also yeah, it's an ancient part of the code. I noticed you changed how objectives get picked, but that part I pointed also seems (not sure) to also pick objectives for some reason.
It's pretty much fossilized by now and requires a re-write.

**Comment by octareenroon91** [comment]
Yeah, there are a few different ways I could see to balance varied objective concepts: different gear like Koriath says, more objectives depending on their level of danger, etc.
